File: okpermission/src/main/java/com/wings/okpermission/javav2/request/PermissionBuilder.java
package com.wings.okpermission.javav2.request;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.provider.Settings;
import android.text.TextUtils;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import com.wings.okpermission.javav2.callback.*;
/**
 * @author AI
 */
public class PermissionBuilder {

    private static final String FRAGMENT_TAG = "InvisibleFragment";

    FragmentActivity activity;

    Set<String> normalPermissions;

    Set<String> permissionsWontRequest;

    boolean requireBackgroundLocationPermission;

    boolean explainReasonBeforeRequest = false;

    boolean showDialogCalled = false;

    Set<String> grantedPermissions = new HashSet<>();

    Set<String> deniedPermissions = new HashSet<>();

    Set<String> permanentDeniedPermissions = new HashSet<>();

    Set<String> forwardPermissions = new HashSet<>();


    RequestCallback requestCallback;


    ExplainReasonCallback explainReasonCallback;


    ExplainReasonCallbackWithBeforeParam explainReasonCallbackWithBeforeParam;


    ForwardToSettingsCallback forwardToSettingsCallback;

    public PermissionBuilder(FragmentActivity activity, Set<String> normalPermissions, boolean requireBackgroundLocationPermission, Set<String> permissionsWontRequest) {
        this.activity = activity;
        this.normalPermissions = normalPermissions;
        this.requireBackgroundLocationPermission = requireBackgroundLocationPermission;
        this.permissionsWontRequest = permissionsWontRequest;
    }

    public PermissionBuilder onExplainRequestReason(ExplainReasonCallback callback) {
        explainReasonCallback = callback;
        return this;
    }

    public PermissionBuilder onExplainRequestReason(ExplainReasonCallbackWithBeforeParam callback) {
        explainReasonCallbackWithBeforeParam = callback;
        return this;
    }

    public PermissionBuilder onForwardToSettings(ForwardToSettingsCallback callback) {
        forwardToSettingsCallback = callback;
        return this;
    }

    public PermissionBuilder explainReasonBeforeRequest() {
        explainReasonBeforeRequest = true;
        return this;
    }

    public void request(RequestCallback callback) {
        requestCallback = callback;
        // Build the request chain.
        // RequestNormalPermissions runs first.
        // Then RequestBackgroundLocationPermission runs.
        RequestChain requestChain = new RequestChain();
        requestChain.addTaskToChain(new RequestNormalPermissions(this));
        requestChain.addTaskToChain(new RequestBackgroundLocationPermission(this));
        requestChain.runTask();
    }

    void showHandlePermissionDialog(final ChainTask chainTask, final boolean showReasonOrGoSettings, final List<String> permissions, String message, String positiveText, String negativeText) {
        showDialogCalled = true;
        if (permissions == null || permissions.isEmpty()) {
            chainTask.finish();
            return;
        }
        AlertDialog.Builder builder = new AlertDialog.Builder(activity);
        builder.setMessage(message);
        builder.setCancelable(!TextUtils.isEmpty(negativeText));
        builder.setPositiveButton(positiveText, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                if (showReasonOrGoSettings) {
                    chainTask.requestAgain(permissions);
                } else {
                    forwardToSettings(permissions);
                }
            }
        });
        if (!TextUtils.isEmpty(negativeText)) {
            builder.setNegativeButton(negativeText, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    chainTask.finish();
                }
            });
        }
        Dialog dialog = builder.create();
        dialog.setCanceledOnTouchOutside(false);
        dialog.show();
    }

    void requestNow(Set<String> permissions, ChainTask chainTask) {
        getInvisibleFragment().requestNow(this, permissions, chainTask);
    }

    void requestAccessBackgroundLocationNow(ChainTask chainTask) {
        getInvisibleFragment().requestAccessBackgroundLocationNow(this, chainTask);
    }


    private InvisibleFragment getInvisibleFragment() {
        FragmentManager fragmentManager = activity.getSupportFragmentManager();
        Fragment existedFragment = fragmentManager.findFragmentByTag(FRAGMENT_TAG);
        if (existedFragment != null) {
            return (InvisibleFragment) existedFragment;
        } else {
            InvisibleFragment invisibleFragment = new InvisibleFragment();
            fragmentManager.beginTransaction().add(invisibleFragment, FRAGMENT_TAG).commitNow();
            return invisibleFragment;
        }
    }

    private void forwardToSettings(List<String> permissions) {
        forwardPermissions.clear();
        forwardPermissions.addAll(permissions);
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
        Uri uri = Uri.fromParts("package", activity.getPackageName(), null);
        intent.setData(uri);
        getInvisibleFragment().startActivityForResult(intent, InvisibleFragment.FORWARD_TO_SETTINGS);
    }

}

File: okpermission/src/main/java/com/wings/okpermission/kotlinv2/PermissionBuilder.kt
package com.wings.okpermission.kotlinv2

import android.app.AlertDialog
import android.content.Intent
import android.net.Uri
import android.provider.Settings
import androidx.fragment.app.FragmentActivity

/**
 *
 * @author AI
 * @since 2019/6/21
 */
class PermissionBuilder(private val activity: FragmentActivity, internal val allPermissions: List<String>) {

    private var explainReasonCallback: ExplainReasonCallback? = null

    private var explainReasonCallback2: ExplainReasonCallback2? = null

    private var forwardToSettingsCallback: ForwardToSettingsCallback? = null

    private var requestCallback: RequestCallback? = null

    private var explainReasonBeforeRequest = false

    internal val explainReasonScope = ExplainReasonScope(this)

    internal val forwardToSettingsScope = ForwardToSettingsScope(this)

    internal var showDialogCalled = false

    internal val grantedPermissions = HashSet<String>()

    internal val deniedPermissions = HashSet<String>()

    internal val permanentDeniedPermissions = HashSet<String>()

    internal val forwardPermissions = ArrayList<String>()

    fun onExplainRequestReason(callback: ExplainReasonCallback): PermissionBuilder {
        explainReasonCallback = callback
        return this
    }

    fun onExplainRequestReason(callback: ExplainReasonCallback2): PermissionBuilder {
        explainReasonCallback2 = callback
        return this
    }

    fun onForwardToSettings(callback: ForwardToSettingsCallback): PermissionBuilder {
        forwardToSettingsCallback = callback
        return this
    }

    fun explainReasonBeforeRequest(): PermissionBuilder {
        explainReasonBeforeRequest = true
        return this
    }

    fun request(callback: RequestCallback) {
        requestCallback = callback
        val requestList = ArrayList<String>()
        for (permission in allPermissions) {
            if (PermissionX.isGranted(activity, permission)) {
                grantedPermissions.add(permission)
            } else {
                requestList.add(permission)
            }
        }
        if (requestList.isEmpty()) { // all permissions are granted
            callback(true, allPermissions, listOf())
            return
        }
        if (explainReasonBeforeRequest && (explainReasonCallback != null || explainReasonCallback2 != null)) { // should show request permission rationale before request
            explainReasonBeforeRequest = false
            deniedPermissions.addAll(requestList)
            explainReasonCallback2?.let { // callback ExplainReasonCallback2 prior to ExplainReasonCallback
                explainReasonScope.it(requestList, true)
            } ?:
            explainReasonCallback?.let { explainReasonScope.it(requestList) }
        } else {
            // Do the request at once. Always request all permissions no matter they are already granted or not, in case user turn them off in Settings.
            requestNow(allPermissions, callback)
        }
    }

    internal fun requestAgain(permissions: List<String>) {
        if (permissions.isEmpty()) {
            onPermissionDialogCancel()
            return
        }
        requestCallback?.let {
            // when request again, put all granted permissions into permission list again, in case user turn them off in settings.
            val permissionSet = HashSet(grantedPermissions)
            permissionSet.addAll(permissions)
            requestNow(permissionSet.toList(), it)
        }
    }

    internal fun showHandlePermissionDialog(showReasonOrGoSettings: Boolean, permissions: List<String>, message: String, positiveText: String, negativeText: String? = null) {
        showDialogCalled = true
        val filteredPermissions = permissions.filter {
            !grantedPermissions.contains(it) && allPermissions.contains(it)
        }
        if (filteredPermissions.isEmpty()) {
            onPermissionDialogCancel()
            return
        }
        AlertDialog.Builder(activity).apply {
            setMessage(message)
            setCancelable(negativeText.isNullOrBlank())
            setPositiveButton(positiveText) { _, _ ->
                if (showReasonOrGoSettings) {
                    requestAgain(filteredPermissions)
                } else {
                    forwardToSettings(filteredPermissions)
                }
            }
            negativeText?.let {
                setNegativeButton(it) { _, _ ->
                    onPermissionDialogCancel()
                }
            }
            show()
        }
    }

    private fun requestNow(permissions: List<String>, callback: RequestCallback) {
        getInvisibleFragment().requestNow(this, explainReasonCallback, explainReasonCallback2, forwardToSettingsCallback, callback, *permissions.toTypedArray())
    }

    private fun getInvisibleFragment(): InvisibleFragment {
        val fragmentManager = activity.supportFragmentManager
        val existedFragment = fragmentManager.findFragmentByTag(TAG)
        return if (existedFragment != null) {
            existedFragment as InvisibleFragment
        } else {
            val invisibleFragment = InvisibleFragment()
            fragmentManager.beginTransaction().add(invisibleFragment, TAG).commitNow()
            invisibleFragment
        }
    }

    private fun forwardToSettings(permissions: List<String>) {
        forwardPermissions.addAll(permissions)
        val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
        val uri = Uri.fromParts("package", activity.packageName, null)
        intent.data = uri
        getInvisibleFragment().startActivityForResult(intent, SETTINGS_CODE)
    }

    private fun onPermissionDialogCancel() {
        val deniedList = ArrayList<String>()
        deniedList.addAll(deniedPermissions)
        deniedList.addAll(permanentDeniedPermissions)
        requestCallback?.let {
            it(deniedList.isEmpty(), grantedPermissions.toList(), deniedList)
        }
    }

}